### ISSUE TITLE
Bitfinex RateLimitExceeded - What is the cause of this restriction?

### ISSUE THREAD

**Issue by KyrMitsos** [ccxt/ccxt#6467]
Hello kroitor,

Thank you for your help to all those people in its need.

I seem to be having the same kind of problems but with Bitfinex now. Would your fix apply to any exchange out there?

If it doesn't then rateLimit doesn't play a role in my case and here is a print out after using --verbose

Object.assign: bitfinex { "error": "ERR_RATE_LIMIT"}
ccxtService.js:274
constructor:class extends BaseClass { … }
message:"bitfinex { "error": "ERR_RATE_LIMIT"}"
name:"RateLimitExceeded"
stack:"RateLimitExceeded: bitfinex { "error": "ERR_RATE_LIMIT"}\n    at bitfinex.throwExactlyMatchedException (c:\Users\<my private path>\node_modules\ccxt\js\base\Exchange.js:533:19)\n    at bitfinex.handleErrors (c:\Users\<my private path>\isonex-fund\coin-price-service\node_modules\ccxt\js\bitfinex.js:1172:22)\n    at response.text.then (c:\Users\<my private path>\node_modules\ccxt\js\base\Exchange.js:628:18)\n    at process._tickCallback (internal/process/next_tick.js:68:7)"
__proto__:Object.assign {constructor: }

I hope I am giving you enough information to understand what this is about.

> What do you think about the cause of this restriction?

The actual reason is that CCXT does not sign public requests for BitMEX, so it didn't work for you with the public `fetchTicker` call. That's an easy fix though. 

Actually, very few exchanges will require authentication (signing) for public requests. And BitMEX `fetchTicker` is a public endpoint: https://www.bitmex.com/api/v1/instrument/activeAndIndices. 

I tried to replicate your error (with public requests unsigned), but in my case it did run for 1 minute or so before breaking down. Looks like their rate limits may vary depending on your location.

Then I tried with signed public calls, and it did work (apparently) – so it continued running without breaking down. Tested it in Python with your snippet and in JS with `node examples/js/cli bitmex fetchTicker BTC/USD --poll --verbose` (with `rateLimit` set to 1100).

I've uploaded the fix, and it will sign the public requests starting from CCXT 1.18.973 (make sure to update your version), if you supply the `apiKey` and `secret` as you do.

Let us know if that does not help. We will be happy if you report back whether it does the trick for you or not. Thx for your involvement!

_Originally posted by @kroitor in https://github.com/ccxt/ccxt/issues/5565#issuecomment-514133560_

**Comment by kroitor** [comment]
@KyrMitsos hi! 

> Would your fix apply to any exchange out there?

Nope, that was BitMEX-specific. Bitfinex and others don't distinguish between authenticated public requests and non-authenticate public requests, therefore Bitfinex handles that a little bit differently.

A quick question to you: do you have the built-in rate-limiter enabled? That is required by the Manual here: https://github.com/ccxt/ccxt/wiki/Manual#rate-limit. If you have it enabled, CCXT should handle the rate-limiting for you. However, if you don't enable this, then you're supposed to rate-limit your requests on the user side. Can you post your exchange instantiation code (with all the options but without your keys of course) ?

**Comment by KyrMitsos** [comment]
Hi Kroitor,

Thank you for the quick reply.

Here is the code you are after:
            let exchange = new ccxt[exchangeName.toLowerCase()]();
            if (exchangeName.toLowerCase() == "bitfinex") {
                exchange.enableRateLimit = true; // enabling ccxt's experimental rate-limiter for 'bitfinex'
                exchange.rateLimit = 10000; // in ms
                exchange.verbose = true;
            }

I looked at the timestamps of the responses and they seem to be 1 to 2 second apart and not roughly 10 seconds that I request above.

Any ideas?

Thank you.

**Comment by kroitor** [comment]
@KyrMitsos 

> I looked at the timestamps of the responses and they seem to be 1 to 2 seconds apart and not roughly 10 seconds that I request above.
> Any ideas?

tbh, that code does not really help, I suggest to do the following (it's a complete program that you can run locally to verify that the rate limiter indeed works and the requests are 10 seconds apart if you set rateLimit = 10000 and enable the built-in rate limiter:

```JavaScript
const ccxt = require ('ccxt')
;(async () => {
    const exchange = new ccxt.bitfinex({
        enableRateLimit: true,
        rateLimit: 10000,
    })
    await exchange.loadMarkets()  // optionally preload them by hand to speed up the first call
    exchange.verbose = true // enable verbosity 
    while (true) {
        console.log (new Date(), 'sending the request...')
        console.log (await exchange.fetchTicker ('BTC/USD'))
    }
}) ()
```

If it does not work – please post your verbose request/response using [github markdown](https://guides.github.com/features/mastering-markdown/) for readability.

**Comment by KyrMitsos** [comment]
Hi Kroitor,

I tried your code and it works like a charm, however looking at my code I cannot understand why that might go wrong, or where.

I have fixed the markdown so please have a look.

Might it be that you are initialising the object with 'new' including the enableRateLimit and rateLimit there?

Looking forward to your reply. :)

Thank you.

**Comment by kroitor** [comment]
> Might it be that you are initialising the object with 'new' including the enableRateLimit and rateLimit as a constructor parameter?

Nope, it should work either way, unless there's a bug somewhere. Here's a slightly edited version of my snippet that is even closer to your code and it creates the exchange from string, then adds the options after initialization:

```JavaScript
const ccxt = require ('ccxt')
    , exchangeName = 'bitfinex'
;(async () => {
    const exchange = new ccxt[exchangeName.toLowerCase ()]
    exchange.enableRateLimit = true
    exchange.rateLimit = 10000
    await exchange.loadMarkets()  // optionally preload them by hand to speed up the first fetchTicker call
    exchange.verbose = true // enable verbosity after loading the markets to avoid too much output
    while (true) {
        console.log (new Date(), 'sending the request...')
        console.log (await exchange.fetchTicker ('BTC/USD'))
    }
}) ()
```

↑ Let me know if it does not work for you. If this snippet works, that means there may be a bug on your side, however, it's hard to tell exactly where, since you haven't pasted all the code to reproduce it, just the initialization bit. Plz try my full snippet above and let me know if that helps.

**Comment by KyrMitsos** [comment]
Hi Kroitor,

It is confirmed. Setting enableRateLimit and rateLimit after the object's constructor initialisation does not have the intented effect (i.e. there is no delay between calls and eventually there is a RATE_LIMIT problem).

Maybe something to either mention in the documentation or fix in a future version.

Either way could you let me know what the ccxt team decides to do. :)

Thank you. You've been very helpful. :)

**Comment by kroitor** [comment]
@KyrMitsos thanks for the feedback! I've uploaded a fix for it and it should work fine as of CCXT version 1.22.48. Let me know if you still have issues with it after upgrading to the most recent version of CCXT. Feel free to reopen this issue or just ask further questions if any. We will be happy if you report back whether it does resolve it for you or not. Thx again!

**Comment by KyrMitsos** [comment]
Thank you Kroitor, I will try the new version and come back if there is a problem.

You've been excellent and I think both myself and others can't thank you enough for your involvement with ccxt and your help. Keep up the good work mate.

**Comment by KyrMitsos** [comment]
Hi Kroitor,

Unfortunately, I keep having these issues on my personal code. I am pasting the Responses from the two setups (the one you have given and the one of my original code):

MY CODE's Response:
```javascript
handleRestResponse:
 bitfinex GET https://api.bitfinex.com/v1/symbols_details 429 Too Many Requests 
Response:
Exchange.js:625
Object {Cache-Control: "no-cache", Cf-Ray: "5651ad24fa5de670-LHR", Connection: "close", Content-Type: "application/json", Date: "Fri, 14 Feb 2020 20:08:50 GMT", …}
Exchange.js:625
Cache-Control:"no-cache"
Cf-Ray:"5651ad24fa5de670-LHR"
Connection:"close"
Content-Type:"application/json"
Date:"Fri, 14 Feb 2020 20:08:50 GMT"
Expect-Ct:"max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct""
Retry-After:"63"
Server:"cloudflare"
Set-Cookie:"__cfduid=dffc17723ba6562ea8206a97b994752121581710930; expires=Sun, 15-Mar-20 20:08:50 GMT; path=/; domain=.bitfinex.com; HttpOnly; SameSite=Lax"
Transfer-Encoding:"chunked"
Vary:"Accept-Encoding"
__proto__:Object {constructor: , __defineGetter__: , __defineSetter__: , …}

 { "error": "ERR_RATE_LIMIT"}
```

YOUR CODE's response:
```javascript
handleRestResponse:
 bitfinex GET https://api.bitfinex.com/v1/pubticker/BTCUSD 200 OK 
Response:
Exchange.js:625
Object {Cache-Control: "max-age=0, private, must-revalidate", Cf-Cache-Status: "DYNAMIC", Cf-Ray: "5651b7021dcf69e1-LHR", Connection: "keep-alive", Content-Encoding: "gzip", …} 
 Cache-Control:"max-age=0, private, must-revalidate"
Cf-Cache-Status:"DYNAMIC"
Cf-Ray:"5651b7021dcf69e1-LHR"
Connection:"keep-alive"
Content-Encoding:"gzip"
Content-Type:"application/json; charset=utf-8"
Date:"Fri, 14 Feb 2020 20:15:34 GMT"
Etag:"W/"394cfb0dbe4b4c3654cbc42b0fa56e4f""
Expect-Ct:"max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct""
Server:"cloudflare"
Set-Cookie:"__cfduid=d77ab2e8f6aeb3992b367a55dc3a18e731581711334; expires=Sun, 15-Mar-20 20:15:34 GMT; path=/; domain=.bitfinex.com; HttpOnly; SameSite=Lax"
Strict-Transport-Security:"max-age=31536000"
Transfer-Encoding:"chunked"
X-Content-Type-Options:"nosniff"
X-Download-Options:"noopen"
X-Frame-Options:"SAMEORIGIN"
X-Permitted-Cross-Domain-Policies:"none"
X-Request-Id:"884473ba-0a7d-4e5c-a8af-2ec7162ceb72"
X-Runtime:"0.004252"
X-Xss-Protection:"1; mode=block"
__proto__:Object {constructor: , __defineGetter__: , __defineSetter__: , …}
```

The only difference I see is the additional
```javascript
Vary:"Accept-Encoding"
```
on my code.

I am sending a bit more code regarding my setup so that you can undestand it a bit better. Actually there is nothing out of the ordinary so here goes:
```javascript
exports.getMarkets = async (exchanges, includeCoins, bothBaseAndQuoteMustIncludeCoins) => {

    const ret = [];

    for (const exchangeName of exchanges) {

        try {
// This I expect to work for all exchanges (not only bitfinex) but used this code to make sure there is no interaction between them
            const exchange = new ccxt[exchangeName.toLowerCase()]({
                    enableRateLimit: true,
                    rateLimit: 6000,
            });
            // const exchange = new ccxt[exchangeName.toLowerCase()]();
            // if (exchangeName.toLowerCase() == "bitfinex") {
            //     exchange.enableRateLimit = true; // enabling ccxt's experimental rate-limiter for 'bitfinex'
            //     exchange.rateLimit = 6100; // in ms
            //     exchange.verbose = true;
            // }
            
            const markets = await exchange.loadMarkets();
.
.
.
```

I used two approaches (the second I left commented) as you can see the function getMarkets takes a list of exchange names as an input parameter and then loops through those calling loadMarkets on each newly created ccxt exchange object.

I have six different occurences of similar code in my app and the operations on each 'exchange' object are:
loadMarkets()
fetchTickers()
fetchCurrencies()

Any ideas? I don't mean to off load this on to you, I am more than happy to work closely to get to the bottom of this. :)

Thank you for your help.

**Comment by KyrMitsos** [comment]
I have done some poking myself. I have changed \ccxt\js\base\Exchange.js, line 506 the function 'fetch2' and entered a couple of console.logs like:
```javascript
    async fetch2 (path, type = 'public', method = 'GET', params = {}, headers = undefined, body = undefined) {
        console.log ("Before " + Date.now());
        if (this.enableRateLimit)
            await this.throttle (this.rateLimit)
        console.log ("After " + Date.now());
        const request = this.sign (path, type, method, params, headers, body)
        return this.fetch (request.url, request.method, request.headers, request.body)
    }
```

On MY CODE I see:
Before 1581765417805
After 1581765417806

On the code you have provided:
Before 1581765681306
After 1581765687247

Clearly on my code's ccxt code the line:
```javascript
await this.throttle (this.rateLimit)
```

Does not work. Why do you think that might be?

The version of ccxt used on both apps is: "^1.22.52" and "1.22.52" respectively (i.e. the same version).

Any ideas as to why that might be happening?

Thank you :)

**Comment by kroitor** [comment]
> Why do you think that might be?

One of the possible reasons is a difference between versions. So you might be using an older version than you think you're using. Let's rule it out first. When pasting code, it has to be complete (remove everything irrelevant from it, leaving just the lines that should definitely work, but don't work for whatever reason), otherwise it's very hard to make sure there's no unwanted interference.

What I posted [here](https://github.com/ccxt/ccxt/issues/6467#issuecomment-585477176) – is a complete short program. Let's make your code complete as well, close to that snippet.

> I have changed \ccxt\js\base\Exchange.js, line 506 the function 'fetch2' and entered a couple of console.logs like

You don't have to add console logging, the same can be enabled by verbose mode:

```JavaScript
'use strict'; // this is important

const ccxt = require ('ccxt')

// print out the runtime version of CCXT explicitly

console.log ('CCXT version:', ccxt.version)

;(async (exchanges) => {

    for (const exchangeId of exchanges) {

        console.log ('--------------------------------------------------------')

        try {

            const exchange = new ccxt[exchangeId]({
                enableRateLimit: true,
                rateLimit: 6000,
            })

            await exchange.loadMarkets ()

            exchange.verbose = true

            // fetch ticker 5 times, for example
            for (let i = 0; i < 5; i++) {
                const index = parseInt (Math.random () * exchange.symbols.length)
                const symbol = exchange.symbols[index]
                console.log (new Date(), exchangeId, 'fetching', symbol)
                console.log (await exchange.fetchTicker (symbol))
            }

        } catch (e) {

            console.log (e)

            // just ignore errors here for now
        }
    }
}) (ccxt.exchanges)
```

↑ Let me know if that works (it can print out some errors for faulty exchanges, but the rate-limiting between fetchTicker requests should enforce 6000 ms delays). I've tested this snippet locally and it works on my side. If the snippet above works (and your runtime CCXT version is indeed up to date), then we will need a similar complete snippet from you in order to find the cause of the issues you're having.

**Comment by KyrMitsos** [comment]
Hi Kroitor,

The snippet you have provided works, but unfortunately not my code.

It is not very easy to post my entire code (or the parts related) because it is an app with 6-7 microservices and one of those (ccxt related) has the issue. There are a couple of API calls that are serviced by that service but unfortunately the code does not block for 6 seconds like the simple example above.

The version I am using is 1.22.52 and that is what comes up when I use this 
```javascript
console.log ('CCXT version:', ccxt.version)
```
on my code.

I don't know how the base code could be different (older) and still get the right kind of version (newer) reported.

Another property of my code is that as I said there is an api call handler which calls an async, which calls another async, which contains the following:
```javascript
console.log ('CCXT version:', ccxt.version);
        
const exchange = new ccxt[exchangeName.toLowerCase()]({
            enableRateLimit: true,
            rateLimit: 6000,
});
var markets = await exchange.loadMarkets();
etc, etc, etc
```
so we have at least 3 awaits in nested function calls.

I don't know if any of this rings any bells. To me the code should work, but for some reason does not block execution when:
```javascript
    async fetch2 (path, type = 'public', method = 'GET', params = {}, headers = undefined, body = undefined) {
        if (this.enableRateLimit)
            await this.throttle (this.rateLimit) // DOES NOT BLOCK HERE
        const request = this.sign (path, type, method, params, headers, body)
        return this.fetch (request.url, request.method, request.headers, request.body)
    }
```

I understand that I am not giving you much to go on, but then again I would need to copy paste a substantial code base to show you the ins and outs and even then you'd require time to figure out the call order etc.

So bottom line, simple code snippets work, but the larger code I am working on doesn't. Is there any chance, or any condition/corner case, at which the await.throttle() would do as expected? I think its that kind of problem we are looking at, like too many nested awaits that just kill it for some reason...

Anyway, thank you for your help. Wish I could be of more assistance. I will keep poking at this, if I have any news I will let you know.

**Comment by kroitor** [comment]
@KyrMitsos it's really hard to help you blindly without seeing at least some reduced form of your code that reproduces the problem in a more obvious way.

> but then again I would need to copy paste a substantial code base to show you the ins and outs and even then you'd require time to figure out the call order etc.

You don't have to paste everything – strip everything that is not related out, leaving just the cascade of three async calls. Or just write a debugging case that triggers the same issue in a different sample. Otherwise you're not helping us to help you )

> so we have at least 3 awaits in nested function calls.

I have tons of that in my code as well as in the examples in this repo – cascades of async calls many levels deep – and that usually works. Have you checked out the examples, btw?

There may be several possible causes, a missing `await` somewhere or some mismanagement of the Promises. Another possible cause is: if you are recreating the exchange instance before each call – that won't work, since the rate-limiter belongs to the exchange instance. There may be other issues as well. 

Sorry, but until you give us something feasible – we can only guess what could be causing it.

**Comment by KyrMitsos** [comment]
Hi Kroitor,

Let me share the crux of how the calls to the ccxt exchange object are done my code:

1) Function A (toUsd) creates the exchange object, and calls loadMarkets(), further down (within it) it calls Function B (getValues). Function B creates the exchange object within it, and then calls loadMarkets(). This call to Function A occurs two times. Each time, the code goes through the string array of exchange names and creates a ccxt exchange object.

2) Function C (getDepositAndWithdrawalFees) creates the exchange object and calls fetchCurrencies(). This call to Function C occurs once for each element of my string array of exchange names.

Functions A and C are orthogonal (unrelated). All functions I describe above are async.

I think what you said about "_Another possible cause is: if you are recreating the exchange instance before each call then that won't work, since the rate-limiter belongs to the exchange instance._" might be close to what is happening, though I would like your opinion on that.

As soon as I find more time, I will try to scale down and mimic the above configuration to a smaller and simpler app (like the one you've given above). This will definitively answer my question, however if you know that the above configuration will not work then I look forward to your feedback and any ways we can avoid this. I will try to think ways around it as well (hopefully without radically restructuring the code), however if this cannot be avoided then that is what I will have to do.

Thank you very much for your time. I hope this is as educational as I find the whole exercise, if not, I hope it is giving something back to you for all your troubles. :)

I am indebted. :)

**Comment by kroitor** [comment]
Hey @KyrMitsos ) The code itself would be a bit more clear than the description of the code, though ))

> might be close to what is happening, though I would like your opinion on that.

From you description it appears that this is exactly the cause of your problem. If your calls are independent, and you are recreating the exchange instance every time – that will effectively reset the rate limiter upon every reinstantiation of the exchange instance.

I suggest to change your logic slightly to reuse the exchange instance if it exists already. That will effectively make your code share the same exchange instance across all methods, regardless of whether they are run in parallel or in a sequence.

1. Create all exchange instances of your interest beforehand at the beginning of your program and place them in a map like so:

```JavaScript
const config = { enableRateLimit: true }
const exchanges = {
    'binance': new ccxt.binance (config),
    'bitfinex': new ccxt.bitfinex (config),
    // ... you can do this in a loop to avoid listing them out by hand
}
```

2. Function A (toUsd) does a lookup in `exchanges` by exchange id and creates the exchange object **if it does not exist in `exchanges`, otherwise it reuses the existing exchange from the map**. And calls loadMarkets(), further down (within it) it calls Function B (getValues). 

3. Function B creates the exchange object within it **if it does not exist in `exchanges`, otherwise it reuses the existing exchange**. And then calls loadMarkets(). This call to Function A occurs two times. Each time, the code goes through the string array of exchange names and creates a ccxt exchange object **if the exchange is not present in `exchanges`, otherwise it reuses the existing exchange instance**.

4. Function C (getDepositAndWithdrawalFees) creates the exchange object **if it does not exist in `exchanges`, otherwise it uses the existing instance** and calls fetchCurrencies(). This call to Function C occurs once for each element of my string array of exchange names.

This should effectively solve the issue by adding the correct architecture to your program.

Let me know if that helps or not.

**Comment by KyrMitsos** [comment]
Yeah, that worked.

Thanks a million Kroitor.

All the best mate :)

**Comment by kroitor** [comment]
@KyrMitsos thx for reporting back! Cheers!
